Create a lightweight SQLite setup and migrate hunk_assingments to use it for persistence

krlvi opened the pull request:
Outstanding issues:
- [x] When a repository is opened for the first time there is a `Failed to run migrations: database is locked` error - likely due to a race condition
    - Byron says: It should be much less likely now, but it's not defeated. It's strange as well.
- [x] Fix CI

Commit:
Setup but-db

Commit:
Initial hunk_assignments table migration

Commit:
Implement a db handle and methods for listing and setting the hunk_assignments table

Commit:
Make the db handle available on the command context

Commit:
Make the hunk assignmetns persistence use the sqlite db instead of toml file

Byron committed:
refactor

* add support for in-memory DBs for testing
* try to fix tests and CI

Byron commented:
I think you can take over now, @krlvi . It will be interesting to see if the added concurrency settings help with that other "Database is locked" issue.

There is still some bits that can still be done if this is still encountered, basically making the DB wait until locks are released instead of failing.

> There are two measures to make this work. Firstly, there should be a busy handler which, in the simplest case, sleeps a little. It should be installed when opening a connection to the database.

https://github.com/crates-io/criner/blob/c36b108c763270a3e5b3adc9c562e0a12646a00d/criner/src/persistence/mod.rs#L59-L60

Secondly each transaction that plans to read before writing should begin in IMMEDIATE mode. Without it, it will first get a read lock for the read portion, and then try to upgrade to a write lock when needed. In my case, this didn't trigger the busy handler and would fail permanently.

> https://github.com/crates-io/criner/blob/c36b108c763270a3e5b3adc9c562e0a12646a00d/criner/src/persistence/table.rs#L135-L160
>
> connection.transaction_with_behavior(rusqlite::TransactionBehavior::Immediate)?
The above is the key to making this work. In conjunction with a busy handler that sleeps…
>
> https://github.com/crates-io/criner/blob/c36b108c763270a3e5b3adc9c562e0a12646a00d/criner/src/persistence/mod.rs#L96-L100
>
> …this call will block until a write lock was obtained before any read happened.

Byron committed:
Configure SQLite for more concurrency.

Prevent DB lock from being fatal by enabling more advanced features, and
by setting up a busy-handler.

However, it's not *really* working.

Byron reviewed (approved):
@krlvi The concurrency issue seems fixed - I wasn't able to reproduce it anymore.
This works thanks to the setup of a busy-handler which waits up to 30s before failing permanently with "database locked".

I would not expect this to be a problem as we only read and write in short bursts, with data (soon) being distributed over multiple tables.

Byron commented on crates/gitbutler-command-context/src/lib.rs:
```diff
@@ -32,6 +35,10 @@ impl CommandContext {
         &self.git_repo
     }
 
+    pub fn db(&mut self) -> &mut but_db::DbHandle {
```
I'd definitely prefer to do this lazily - with `&mut self` that's easy, and won't spawn a database each time `git_head()` is called (this is where it failed for me with "database locked".

Byron committed:
Make the DB lazily initialized to avoid creating it unless needed.

Byron commented:
I also tried this (change order of Pragma): https://github.com/diesel-rs/diesel/issues/2365#issuecomment-2899347817 .
Let's see.
